torch/quantization/fx/quantize.py

```diff
@@ -407,6 +414,687 @@ def load_arg(a):
+    passes which optimize the graph by deduplicating observers, etc.
+    """
+
+    env: Dict[Any, Any] = {}
+    node_name_to_target_dtype: Dict[str, Any] = {}
+    cache_for_no_tensor_check: Dict[Node, bool] = dict()
+
+    def load_arg(a):
+        return map_arg(a, lambda node: env[node.name])
+
+    inputs_seen_counter = 0
+    outputs_seen_counter = 0
+    results_node = None
+
+    # first, populate the dtype map based only on qconfig and qhandler
+    # this assumes:
+    # graph inputs are fp32 by default, and int8 where overriden
+    # other nodes output dtype is specified by the qconfig
+    for node in model.graph.nodes:
+        root_node, matched_nodes, pattern, qhandler, qconfig = matches.get(
+            node.name, (None, None, None, None, None))
+        node_name_to_target_dtype[node.name] = get_target_dtype_for_node_rewrite(
+            node, qconfig, inputs_seen_counter, outputs_seen_counter,
+            input_quantized_idxs, output_quantized_idxs, qhandler)
+
+    # Second, for nodes with known input dtypes, propagate them throughout the
+    # graph. For example, if there is a call such as
+    #   x1 = x0.masked_fill(mask, 1)
+    # we propagate the type of mask to be torch.bool
+    propagate_dtypes_for_known_nodes_rewrite(
+        model.graph, node_name_to_target_dtype, matches)
+
+    for node in model.graph.nodes:
+        # print('\nobs graph', observed_graph, '\n')
+        # print('\nenv', env)
+        # print('\nnode_name_to_target_dtype', node_name_to_target_dtype)
+        # print('processing node', node.format_node(), '\n')
+
+        # copy the current node as is
+        new_node = observed_graph.node_copy(node, load_arg)
+        env[node.name] = new_node
+
+        # check for matches
+        root_node, matched_nodes, pattern, qhandler, qconfig = matches.get(
+            node.name, (None, None, None, None, None))
+        # print('match results', root_node, matched_nodes, pattern, qhandler, qconfig)
+
+        #
+        # After this point, the current node and all of its arguments
+        # have a dtype assigned. Now, we insert observers for inputs
+        # of this node (if needed for this node), and the output of this node
+        # (if needed for this node).
+        #
+
+        if new_node.op == 'placeholder':
+            # if a graph input is in fp32, it does not need observation
+            # if a graph input is in int8, we assume the observation happens
+            #   outside of the graph, and no additional observation is needed
+            pass
+
+        elif new_node.op in ('call_module', 'call_method', 'call_function', 'output'):
+            modules = dict(model.named_modules(remove_duplicate=False))
+            args_have_no_tensors = \
+                all_node_args_have_no_tensors(
+                    new_node, modules, cache_for_no_tensor_check)
+            is_getitem = new_node.op == 'call_function' and \
```

Comment:
this would be equivalent to removing operator.getitem in the quantization_patterns.py? does the test_general_shape_ops work?

Reply:
the `test_general_shape_ops` test passes.  Removing operator.getitem from patterns sounds good to me, maybe that can be done in a future PR, just to keep this PR as a refactor with minimal logic changes?